File: linklist/21.cpp
#include<iostream>
using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
        ListNode head(0);
        ListNode* tmp = &head;
        
        while(l1 && l2){
            if(l1->val > l2->val){
                tmp->next = l2;
                l2 = l2->next;
            }else{
                tmp->next = l1;
                l1 = l1->next;
            }
            tmp = tmp->next;
        }
        
        tmp->next = l1?l1:l2;
        return head.next;

    }
};
class Solution2 {
public:
    ListNode *mergeTwoLists(ListNode *l1, ListNode *l2) {
        ListNode dummy(INT_MIN);
        ListNode *tail = &dummy;
        
        while (l1 && l2) {
            if (l1->val < l2->val) {
                tail->next = l1;
                l1 = l1->next;
            } else {
                tail->next = l2;
                l2 = l2->next;
            }
            tail = tail->next;
        }

        tail->next = l1 ? l1 : l2;
        return dummy.next;
    }
};
int main(){
    ListNode head(0);
    ListNode one(1);
    ListNode two(0);
    ListNode three(0);
    head.next = &one;
    two.next = &three;

    Solution s;
    ListNode* result = s.mergeTwoLists(&head, &two);
    while(result!=NULL){
        cout<<result->val<<endl;
        result = result->next;
    }
    return 0;
}

File: linklist/21.py
# coding=UTF-8
'''
@Author: httermin
@Date: 2019-12-30 15:48:07
'''
import sys
import os

sys.path.append("..")
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def mergeTwoLists(self, l1: ListNode, l2: ListNode) -> ListNode:
        if l1 is None: return l2
        if l2 is None: return l1

        if l1.val > l2.val:
            start = l2
            l2 = l2.next
        else:
            start = l1
            l1 = l1.next
        
        # l1 is the smaller one

        tmp = start
        
        while (l1 is not None) or (l2 is not None):
            
            if l1 is None:
                tmp.next = l2
                l2 = l2.next
            elif l2 is None:
                tmp.next = l1
                l1 = l1.next
            else:
                if l1.val < l2.val:
                    tmp.next = l1
                    l1 = l1.next
                else:
                    tmp.next = l2
                    l2 = l2.next
            tmp = tmp.next
        return start
        